### ISSUE TITLE
Memory leak using loss in Eager Execution

### ISSUE THREAD

**Issue by henriquem87** [tensorflow/tensorflow#20062]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Xubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: v1.8.0-3211-g1aea422 1.9.0-rc0
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**: 0.11.1
- **GCC/Compiler version (if compiling from source)**: 5.4.0
- **CUDA/cuDNN version**: 9.1 / 7.1
- **GPU model and memory**: GeForce GTX 960M - 2GB
- **Exact command to reproduce**: The following code:

```python
import tensorflow as tf
tf.enable_eager_execution()

layer = tf.keras.layers.Dense(5)
inputs = tf.zeros([32, 100], tf.float32)
labels = tf.zeros([32], tf.int64)

for i in range(100000):
  with tf.GradientTape() as tape:
    logits = layer(inputs)
    loss = tf.losses.sparse_softmax_cross_entropy(labels, logits)
```

### Describe the problem

I observe a memory leak when computing the loss using the code above. I also tried using other losses, such as `tf.losses.mean_squared_error` and `tf.losses.hinge_loss`, and I observed similar leaks. Using memory_profiler produces the following plot:
![mplot_leak_entropy](https://user-images.githubusercontent.com/24420973/41473326-53606eb2-708f-11e8-9618-2f7696781828.png)

I tried to use the workaround suggested by @allenlavoie in #19671, but it does not seem to work in this case.

### Source code / logs

I used pympler to track the memory usage, by adding it to the code:

```python
import tensorflow as tf
from pympler.tracker import SummaryTracker
tf.enable_eager_execution()

layer = tf.keras.layers.Dense(5)
inputs = tf.zeros([32, 100], tf.float32)
labels = tf.zeros([32], tf.int64)

tracker = SummaryTracker()

for i in range(100000):
  with tf.GradientTape() as tape:
    logits = layer(inputs)
    loss = tf.losses.sparse_softmax_cross_entropy(labels, logits)
  if i % 1000 == 0:
    tracker.print_diff()
```

After a few rounds, it starts to produce the following output:
```bash
                types |   # objects |   total size
===================== | =========== | ============
  <class 'EagerTensor |        1000 |    156.25 KB
         <class 'list |           0 |      7.14 KB
```

So it seems an EagerTensor is being created and not released at every iteration, which may be the cause of the leak.

I thank @allenlavoie and @akshaym for solving my previous issues in #19385. I think this problem is related to those.

**Comment by allenlavoie** [comment]
I see, this one is much simpler:

```
import tensorflow as tf
tf.enable_eager_execution()

layer = tf.keras.layers.Dense(5)
inputs = tf.zeros([32, 100], tf.float32)
labels = tf.zeros([32], tf.int64)

for i in range(100000):
  tf.losses.sparse_softmax_cross_entropy(labels, inputs)
  print(len(tf.get_default_graph().get_collection(tf.GraphKeys.LOSSES)))
```

You can `tf.reset_default_graph()` to work around the issue. The fix will probably be to not add to a global collection. Thank you for the report!

**Comment by henriquem87** [comment]
Thank you for the reply @allenlavoie . This workaround fixes the problem for the loss, but it seems to create a leak with `tf.contrib.summary.record_summaries_every_n_global_steps`. This code:
```python
import tensorflow as tf
tf.enable_eager_execution()

for i in range(100000):
  tf.reset_default_graph() 
  with tf.contrib.summary.record_summaries_every_n_global_steps(100):
    pass
```
produces the following memory plot:
![mprof_summary_leak](https://user-images.githubusercontent.com/24420973/41495224-d8ae5a7a-70f8-11e8-88e0-0c22ef9c2c00.png)

Strangely, this leak only happens if we reset the graph. If we remove `tf.reset_default_graph()` from the loop, there is no leak at all. Removing the summary from the loop also stops the leak, so it is not the reset operation which leaks.

**Comment by allenlavoie** [comment]
Summaries may be another case related to https://github.com/tensorflow/tensorflow/issues/19671. I think they both create resources with unique names, so they'll both hit our kernel cache. You could probably reset the graph and set the random seed to work around them both.

I am planning to fix the kernel caching issues (hopefully relatively soon), and I have a fix out for review on the losses collection issue. Thank you for the reports!
